build_tools/cmake/iree_run_module_test.cmake

```diff
@@ -0,0 +1,369 @@
+    ${ARGN}
+  )
+
+  if(CMAKE_CROSSCOMPILING AND "hostonly" IN_LIST _RULE_LABELS)
+    return()
+  endif()
+
+  iree_get_platform(_PLATFORM)
+  if(_PLATFORM IN_LIST _RULE_UNSUPPORTED_PLATFORMS)
+    return()
+  endif()
+
+  set(_RUNNER_TARGET "iree-run-module")
+  iree_package_name(_PACKAGE_NAME)
+  set(_NAME "${_PACKAGE_NAME}_${_RULE_NAME}")
+
+  get_filename_component(_SRC "${_RULE_MODULE_SRC}" REALPATH)
+  list(APPEND _RULE_RUNNER_ARGS "--module_file=${_SRC}")
+
+  if(_RULE_EXPECTED_OUTPUT)
+    get_filename_component(_OUTPUT_FILE_TYPE "${_RULE_EXPECTED_OUTPUT}" EXT)
+    if(_OUTPUT_FILE_TYPE STREQUAL ".txt")
+      get_filename_component(_OUTPUT_FILE_SRC "${_RULE_EXPECTED_OUTPUT}" REALPATH)
+      # Process the text input to remove the line breaks.
+      file(STRINGS "${_OUTPUT_FILE_SRC}" _EXPECTED_OUTPUT ENCODING UTF-8)
+      string(REPLACE ";" " " _EXPECTED_OUTPUT_STR "${_EXPECTED_OUTPUT}")
+      set(_EXPECTED_OUTPUT_STR "--expected_output=\"${_EXPECTED_OUTPUT_STR}\"")
+      list(APPEND _RULE_RUNNER_ARGS ${_EXPECTED_OUTPUT_STR})
+      list(APPEND _SRC ${_OUTPUT_FILE_SRC})
+    elseif(_OUTPUT_FILE_TYPE STREQUAL ".npy")
+      # Large npy files are not stored in the codebase. Need to download them
+      # from GCS iree-model-artifacts first and store them in the following possible
+      # paths.
+      find_file(_OUTPUT_FILE_SRC_PATH
+        NAME
+          "${_RULE_EXPECTED_OUTPUT}"
+        PATHS
+          "${CMAKE_CURRENT_SOURCE_DIR}"
+          "${CMAKE_CURRENT_BINARY_DIR}"
+          "${IREE_BENCHMARK_SUITE_DIR}"
+        NO_CACHE
+        NO_DEFAULT_PATH
+      )
+      # If the expected output npy file is not found (the large file is not
+      # loaded from GCS to `IREE_BENCHMARK_SUITE_DIR` benchmark suite test),
+      # report error.
+      if(NOT _OUTPUT_FILE_SRC_PATH)
+        message(SEND_ERROR "${_RULE_EXPECTED_OUTPUT} is not found in\n\
+          ${CMAKE_CURRENT_SOURCE_DIR}\n\
+          ${CMAKE_CURRENT_BINARY_DIR}\n\
+          ${IREE_BENCHMARK_SUITE_DIR}\n\
+          Please check if you need to download it first.")
+      else()
+        get_filename_component(_OUTPUT_FILE_SRC "${_OUTPUT_FILE_SRC_PATH}" REALPATH)
+        list(APPEND _RULE_RUNNER_ARGS "--expected_output=@${_OUTPUT_FILE_SRC}")
+        list(APPEND _SRC ${_OUTPUT_FILE_SRC})
+      endif()
+    else()  # The expected output is listed in the field.
+      list(APPEND _RULE_RUNNER_ARGS "--expected_output=\"${_RULE_EXPECTED_OUTPUT}\"")
+    endif()
+  endif(_RULE_EXPECTED_OUTPUT)
+
+  if(NOT DEFINED _RULE_DRIVER)
+    return()
+  endif()
+
+  # Dump the flags into a flag file to avoid CMake's naive handling of spaces
+  # in expected output.
+  if(_RULE_RUNNER_ARGS)
+    set(_OUTPUT_FLAGFILE "${CMAKE_CURRENT_BINARY_DIR}/${_RULE_NAME}_flagfile")
+    # Write each argument in a new line.
+    string(REPLACE ";" "\n" _OUTPUT_FLAGS "${_RULE_RUNNER_ARGS}")
+    file(WRITE "${_OUTPUT_FLAGFILE}" "${_OUTPUT_FLAGS}")
```

Comment:
I think `file(CONFIGURE OUTPUT ...)` or `file(GENERATE OUTPUT ...)` would be better here. These will only regenerate the file when the content changes and therefore will be properly integrated with the build system, etc. `file(write ...)` is basically creating a side-effect, whereas these commands are creating a generated file

https://cmake.org/cmake/help/latest/command/file.html#configure
https://cmake.org/cmake/help/latest/command/file.html#generate

This would also allow making a template string rather than some of the string manipulation you're doing, which would perhaps be preferable (although, then again, maybe having all the args in a list is nice)